### ISSUE TITLE
Fail to build the stack

### ISSUE THREAD

**Issue by lizzzcai** [OpenFunction/builder#55]
I tried to follow [docs](https://github.com/OpenFunction/builder/tree/main/builders/go116#build-go116-stack) to build the `go116` stack and builder, but facing some errors.

```
❯ pwd
/Users/home/dev/openfunction/builder
❯ bazel run //builders/go116/stack:build
INFO: Analyzed target //builders/go116/stack:build (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //builders/go116/stack:build up-to-date:
  bazel-bin/builders/go116/stack/build
INFO: Elapsed time: 0.478s, Critical Path: 0.01s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
> Extracting licenses tar
> Building base go116common image
[+] Building 3.3s (9/9) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                   0.0s
 => => transferring dockerfile: 1.49kB                                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/openfunctiondev/ubuntu_18_0_4:latest                                                                                                                                        3.1s
 => [auth] openfunctiondev/ubuntu_18_0_4:pull token for registry-1.docker.io                                                                                                                                           0.0s
 => [1/4] FROM docker.io/openfunctiondev/ubuntu_18_0_4@sha256:47202afe7bd8fb1eebfb6175f5eb4afba6747c22599dc04b53689574bb0a4f3b                                                                                         0.0s
 => CACHED [2/4] RUN apt-get update && apt-get install -y --no-install-recommends   libexpat1   libffi6   libmpdec2   libicu60   libc++1-9   tzdata   && apt-get clean && rm -rf /var/lib/apt/lists/*                  0.0s
 => CACHED [3/4] RUN groupadd cnb --gid 1000 &&   useradd --uid 1000 --gid 1000 -m -s /bin/bash cnb                                                                                                                    0.0s
 => CACHED [4/4] RUN curl --fail --show-error --silent --location --retry 3 https://go.dev/dl/go1.16.13.linux-amd64.tar.gz | tar xz --directory /usr/local/ --strip-components=1                                       0.0s
 => exporting to image                                                                                                                                                                                                 0.0s
 => => exporting layers                                                                                                                                                                                                0.0s
 => => writing image sha256:d72db7c49a80d4c940f335b01f31f1b608b2a6657a7bf9eda958f017f2862082                                                                                                                           0.0s
 => => naming to docker.io/library/go116common                                                                                                                                                                         0.0s
> Building base go116run image
[+] Building 2.9s (7/7) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                   0.0s
 => => transferring dockerfile: 865B                                                                                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/library/busybox:1.35                                                                                                                                                        2.8s
 => [auth] library/busybox:pull token for registry-1.docker.io                                                                                                                                                         0.0s
 => [1/2] FROM docker.io/library/busybox:1.35@sha256:0d1d11dbad88f301fe745cf1b3fca8ee44ff5cf1c1be721f36b22bec592d9966                                                                                                  0.0s
 => CACHED [2/2] RUN adduser -u 1000 -D cnb cnb                                                                                                                                                                        0.0s
 => exporting to image                                                                                                                                                                                                 0.0s
 => => exporting layers                                                                                                                                                                                                0.0s
 => => writing image sha256:f8f8e11174a63f35c447d5ba415a1b3095f5ad75734e38cf76ec80cd8fdc7f1b                                                                                                                           0.0s
 => => naming to docker.io/library/go116run                                                                                                                                                                            0.0s
> Building openfunctiondev/buildpacks-run-go:v2
[+] Building 0.1s (5/5) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                   0.0s
 => => transferring dockerfile: 678B                                                                                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/library/go116run:latest                                                                                                                                                     0.0s
 => CACHED [1/1] FROM docker.io/library/go116run                                                                                                                                                                       0.0s
 => exporting to image                                                                                                                                                                                                 0.0s
 => => exporting layers                                                                                                                                                                                                0.0s
 => => writing image sha256:9a946a62545b94c565c12b0414ff6ba9647ecf001be8623a182d363f94376477                                                                                                                           0.0s
 => => naming to docker.io/openfunctiondev/buildpacks-run-go:v2                                                                                                                                                        0.0s
> Building openfunctiondev/buildpacks-go116-build:v2
[+] Building 0.1s (1/2)
 => [internal] load build definition from build.Dockerfile                                                                                                                                                             0.0s
 => => transferring dockerfile: 121B                                                                                                                                                                                   0.0s
failed to solve with frontend dockerfile.v0: failed to read dockerfile: open /var/lib/docker/tmp/buildkit-mount085995818/Users/home/dev/openfunction/builder/builders/go116/stack/build.Dockerfile: no such file or directory
```

Command failed is:

```
docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" -f "${DIR}/build.Dockerfile" "${TEMP}"
```

if I removed the `"${TEMP}"`, the Dockerfile can be found, but facing another error:

```
❯ bazel run //builders/go116/stack:build
INFO: Analyzed target //builders/go116/stack:build (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //builders/go116/stack:build up-to-date:
  bazel-bin/builders/go116/stack/build
INFO: Elapsed time: 0.365s, Critical Path: 0.01s
INFO: 2 processes: 2 internal.
INFO: Build completed successfully, 2 total actions
INFO: Build completed successfully, 2 total actions
> Extracting licenses tar
> Building base go116common image
[+] Building 1.7s (8/8) FINISHED                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                  0.0s
 => => transferring dockerfile: 1.49kB                                                                                                                                                0.0s
 => [internal] load .dockerignore                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/openfunctiondev/ubuntu_18_0_4:latest                                                                                                       1.6s
 => [1/4] FROM docker.io/openfunctiondev/ubuntu_18_0_4@sha256:47202afe7bd8fb1eebfb6175f5eb4afba6747c22599dc04b53689574bb0a4f3b                                                        0.0s
 => CACHED [2/4] RUN apt-get update && apt-get install -y --no-install-recommends   libexpat1   libffi6   libmpdec2   libicu60   libc++1-9   tzdata   && apt-get clean && rm -rf /va  0.0s
 => CACHED [3/4] RUN groupadd cnb --gid 1000 &&   useradd --uid 1000 --gid 1000 -m -s /bin/bash cnb                                                                                   0.0s
 => CACHED [4/4] RUN curl --fail --show-error --silent --location --retry 3 https://go.dev/dl/go1.16.13.linux-amd64.tar.gz | tar xz --directory /usr/local/ --strip-components=1      0.0s
 => exporting to image                                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                                               0.0s
 => => writing image sha256:d72db7c49a80d4c940f335b01f31f1b608b2a6657a7bf9eda958f017f2862082                                                                                          0.0s
 => => naming to docker.io/library/go116common                                                                                                                                        0.0s
> Building base go116run image
[+] Building 1.6s (6/6) FINISHED                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                  0.0s
 => => transferring dockerfile: 865B                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/busybox:1.35                                                                                                                       1.5s
 => [1/2] FROM docker.io/library/busybox:1.35@sha256:0d1d11dbad88f301fe745cf1b3fca8ee44ff5cf1c1be721f36b22bec592d9966                                                                 0.0s
 => CACHED [2/2] RUN adduser -u 1000 -D cnb cnb                                                                                                                                       0.0s
 => exporting to image                                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                                               0.0s
 => => writing image sha256:f8f8e11174a63f35c447d5ba415a1b3095f5ad75734e38cf76ec80cd8fdc7f1b                                                                                          0.0s
 => => naming to docker.io/library/go116run                                                                                                                                           0.0s
> Building openfunctiondev/buildpacks-run-go:v2
[+] Building 0.2s (5/5) FINISHED                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                  0.0s
 => => transferring dockerfile: 678B                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/go116run:latest                                                                                                                    0.0s
 => CACHED [1/1] FROM docker.io/library/go116run                                                                                                                                      0.0s
 => exporting to image                                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                                               0.0s
 => => writing image sha256:9a946a62545b94c565c12b0414ff6ba9647ecf001be8623a182d363f94376477                                                                                          0.0s
 => => naming to docker.io/openfunctiondev/buildpacks-run-go:v2                                                                                                                       0.0s
> Building openfunctiondev/buildpacks-go116-build:v2
"docker build" requires exactly 1 argument.
See 'docker build --help'.

Usage:  docker build [OPTIONS] PATH | URL | -

Build an image from a Dockerfile
```

System: MacOS 12.3.1
```
❯ docker --version
Docker version 20.10.14, build a224086
❯ bazel --version
bazel 5.1.1-homebrew
```

**Comment by benjaminhuo** [comment]
It's working on Linux env.
You can change 
```shell
docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" -f "${DIR}/build.Dockerfile" "${TEMP}"
```

to 
```shell
docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" 
 - < "${DIR}/build.Dockerfile"
```

**Comment by lizzzcai** [comment]
> It's working on Linux env. You can change
> 
> ```shell
> docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" -f "${DIR}/build.Dockerfile" "${TEMP}"
> ```
> 
> to
> 
> ```shell
> docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" 
>  - < "${DIR}/build.Dockerfile"
> ```


removed `"${TEMP}"` will cause a issue, as there is a licenses in the `"${TEMP}"`. And the build.Dockerfile is trying to copy it.

```
ARG from_image
FROM ${from_image}
COPY licenses/ /usr/local/share/licenses/buildpacks/
...
```

```
❯ bazel run //builders/go116/stack:build
INFO: Analyzed target //builders/go116/stack:build (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //builders/go116/stack:build up-to-date:
  bazel-bin/builders/go116/stack/build
INFO: Elapsed time: 0.416s, Critical Path: 0.02s
INFO: 2 processes: 2 internal.
INFO: Build completed successfully, 2 total actions
INFO: Build completed successfully, 2 total actions
> Extracting licenses tar
> Building base go116common image
[+] Building 1.5s (8/8) FINISHED                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                  0.0s
 => => transferring dockerfile: 1.49kB                                                                                                                                                0.0s
 => [internal] load .dockerignore                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/openfunctiondev/ubuntu_18_0_4:latest                                                                                                       1.4s
 => [1/4] FROM docker.io/openfunctiondev/ubuntu_18_0_4@sha256:47202afe7bd8fb1eebfb6175f5eb4afba6747c22599dc04b53689574bb0a4f3b                                                        0.0s
 => CACHED [2/4] RUN apt-get update && apt-get install -y --no-install-recommends   libexpat1   libffi6   libmpdec2   libicu60   libc++1-9   tzdata   && apt-get clean && rm -rf /va  0.0s
 => CACHED [3/4] RUN groupadd cnb --gid 1000 &&   useradd --uid 1000 --gid 1000 -m -s /bin/bash cnb                                                                                   0.0s
 => CACHED [4/4] RUN curl --fail --show-error --silent --location --retry 3 https://go.dev/dl/go1.16.13.linux-amd64.tar.gz | tar xz --directory /usr/local/ --strip-components=1      0.0s
 => exporting to image                                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                                               0.0s
 => => writing image sha256:d72db7c49a80d4c940f335b01f31f1b608b2a6657a7bf9eda958f017f2862082                                                                                          0.0s
 => => naming to docker.io/library/go116common                                                                                                                                        0.0s
> Building base go116run image
[+] Building 1.2s (6/6) FINISHED                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                  0.0s
 => => transferring dockerfile: 865B                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/busybox:1.35                                                                                                                       1.1s
 => [1/2] FROM docker.io/library/busybox:1.35@sha256:0d1d11dbad88f301fe745cf1b3fca8ee44ff5cf1c1be721f36b22bec592d9966                                                                 0.0s
 => CACHED [2/2] RUN adduser -u 1000 -D cnb cnb                                                                                                                                       0.0s
 => exporting to image                                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                                               0.0s
 => => writing image sha256:f8f8e11174a63f35c447d5ba415a1b3095f5ad75734e38cf76ec80cd8fdc7f1b                                                                                          0.0s
 => => naming to docker.io/library/go116run                                                                                                                                           0.0s
> Building openfunctiondev/buildpacks-run-go:v2
[+] Building 0.1s (5/5) FINISHED                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                  0.0s
 => => transferring dockerfile: 678B                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/go116run:latest                                                                                                                    0.0s
 => CACHED [1/1] FROM docker.io/library/go116run                                                                                                                                      0.0s
 => exporting to image                                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                                               0.0s
 => => writing image sha256:9a946a62545b94c565c12b0414ff6ba9647ecf001be8623a182d363f94376477                                                                                          0.0s
 => => naming to docker.io/openfunctiondev/buildpacks-run-go:v2                                                                                                                       0.0s
> Building openfunctiondev/buildpacks-go116-build:v2
[+] Building 0.1s (6/7)                                                                                                                                                                    
 => [internal] load build definition from Dockerfile                                                                                                                                  0.0s
 => => transferring dockerfile: 1.23kB                                                                                                                                                0.0s
 => [internal] load .dockerignore                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/go116common:latest                                                                                                                 0.0s
 => [internal] load build context                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                       0.0s
 => CACHED [1/3] FROM docker.io/library/go116common                                                                                                                                   0.0s
 => ERROR [2/3] COPY licenses/ /usr/local/share/licenses/buildpacks/                                                                                                                  0.0s
------
 > [2/3] COPY licenses/ /usr/local/share/licenses/buildpacks/:
------
failed to compute cache key: "/licenses" not found: not found
```

**Comment by benjaminhuo** [comment]
```
docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" "${TEMP}"
 - < "${DIR}/build.Dockerfile"
```

**Comment by lizzzcai** [comment]
> ```
> docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" "${TEMP}"
>  - < "${DIR}/build.Dockerfile"
> ```

Hi @benjaminhuo , thanks for the clue. This one still not working in MacOS, but I finally fix it by:
```
cat "${DIR}/build.Dockerfile" | docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" -f - "${TEMP}"
```
I will improve it together in my upcoming PR. This issue can be closed now.

**Comment by webup** [comment]
> > ```
> > docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" "${TEMP}"
> >  - < "${DIR}/build.Dockerfile"
> > ```
> 
> Hi @benjaminhuo , thanks for the clue. This one still not working in MacOS, but I finally fix it by:
> 
> ```
> cat "${DIR}/build.Dockerfile" | docker build --build-arg "from_image=go116common" -t "openfunctiondev/buildpacks-go116-build:$TAG" -f - "${TEMP}"
> ```
> 
> I will improve it together in my upcoming PR. This issue can be closed now.

It should be better to close this issue with your PR, with some issue closing keywords embedded. :)

**Comment by lizzzcai** [comment]
closed by #56
